### PR TITLE
Added Sky Drop to Battle Engine

### PR DESCRIPTION
**Pull request by WiserVisor**
Added in the move Sky Drop, which came with some new macros, methods, and memory allocations to get it to work.
Second pull request, due to messy commits and some unnecessary changes within first pull request.

## Description
Sky Drop works just as it does from Gen 6 onwards.

When a Pokemon is released from Sky Drop, the text: "{TARGET} was released from Sky Drop!" doesn't appear. This is less of an issue and more the fact that I found that bit of text unnecessary in the games.

## **Discord contact info**
N/A
Contact me on PokeCommunity
WiserVisor https://www.pokecommunity.com/member.php?u=682852

### CODE REVIEW

File: include/constants/battle.h
Comment (AsparagusEduardo) [reply include/constants/battle.h:174]: Once #1742 is merged, `STATUS3_SKY_DROPPED` can be its own flag.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
There should be a check in `CanMegaEvolve` that prevents the user from Mega Evolving if it's being Sky Dropped.

Tested a couple things against Gen 5 (sorry for the bad gifs):
- Sky Drop should fail when aimed to allies:
SD_allies
SDG5_allies
- If the target of Sky Drop faints before taking damage, it should then fail instead of switching damage to the other mon.
SD_changeSide
SDG5_changeSide
- Instead of the Flying-type mon taking 0 damage, it should be freed and the message "It doesn't affect X" should show.
SD_flyingtype
)

File: data/maps/LittlerootTown_ProfessorBirchsLab/scripts.inc
Comment (LOuroboros) [inline data/maps/LittlerootTown_ProfessorBirchsLab/scripts.inc]:
```suggestion

```

File: data/battle_scripts_1.s
Comment (LOuroboros) [inline data/battle_scripts_1.s]: Won't removing this make it so Sky Drop will no longer print the "Pokémon used Sky Drop!" string?

Comment (WiserVisor) [reply data/battle_scripts_1.s]: No, because attackstring is used in BattleScriptFirstChargingTurn.

File: data/battle_scripts_1.s
Comment (LOuroboros) [inline data/battle_scripts_1.s:440]: Tab instead of spaces.

File: src/battle_script_commands.c
Comment (LOuroboros) [inline src/battle_script_commands.c:8880]: Spaces instead of tabs.

File: src/battle_script_commands.c
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:8890]: I've been thinking... Mon shouldn't be able to switch while Skydropping or being skydropped, so why is it saving `gBattlerPartyIndexes` instead of battlerIds?

Comment (WiserVisor) [reply src/battle_script_commands.c:8890]: Because the mon might faint in the air while it is being sky dropped. If battlerIds were used instead, the following error could (and did) occur: It's a trainer double battle. Pokémon A targets Pokémon B with Sky Drop and takes B into the air. Pokémon C targets B with a hit_in_air move, causing Pokémon B to faint. The trainer of B sends out Pokémon D to replace B. If battlerIds are used, Sky Drop will still deal damage to D, even though the original target of the move was B, since B and D would end up having the same battlerId. Using the party index is the only way to be absolutely sure that the target of Sky Drop on the first turn is still the same target on the second turn.

Comment (AsparagusEduardo) [reply src/battle_script_commands.c:8908]: Alright, that makes sense.
One more thing before I approve: When using `gBattlerPartyIndexes` why is it using +- 6 when saving and reading? Wouldn't it be simpler to just save and use the party index without that modification?

Comment (ghoulslash) [reply src/battle_script_commands.c:8908]: I think it would be much cleaner to initialize `skyDropTargets` as `0xFF` in `BattleStartSetClearData`, and reset the field for `gActiveBattler` to 0xFF in `SwitchInClearSetData` and `FaintClearSetData`. Then you won't need the +- 6 checks which are very difficult to read/comprehend, and you can go ahead and say the attack turn of sky drop fails if the skydroptarget is 0xFF

Comment (WiserVisor) [reply src/battle_script_commands.c]: I did the +-6 thing since I had the cleared value for skyDropTargets as 0, but 0 is a valid index for a party index, so I had to offset it by 6 to make sure there was no overlap over the valid values. Switching the cleared value to 0xFF as ghoulslash suggested made that unnecessary and made a lot of things much simpler besides.

**Comment by LOuroboros** [comment]
Not gonna lie, the commit history of this PR is an absolute mess. It's probably the biggest amount of commits in a single PR submitted to this project 🤣 ...

I think it would be a good idea to move this to a clean copy of the BE and open a new PR with a single commit. Specially since the changes don't seem to be particularly big or tedious to reinsert.

**Pull request closed without merging**
